File: 03. Stack and Queue/04. Previous Greater Element/main.java
import java.util.Stack;

public class main {
    public static int[] previousGreaterElement(int[] nums) {
        Stack<Integer> stack = new Stack<>();
        int[] result = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            while (!stack.isEmpty() && stack.peek() <= nums[i]) {
                stack.pop();
            }
            result[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(nums[i]);
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums = {4, 5, 2, 10, 8};
        int[] result = previousGreaterElement(nums);

        System.out.println("Previous Greater Elements:");
        for (int num : result) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}

File: 03. Stack and Queue/04. Previous Greater Element/main.py
def previous_greater_element(nums):
    stack = []
    result = [-1] * len(nums)

    for i in range(len(nums)):
        while stack and stack[-1] <= nums[i]:
            stack.pop()
        if stack:
            result[i] = stack[-1]
        stack.append(nums[i])

    return result

def main():
    nums = [4, 5, 2, 10, 8]
    result = previous_greater_element(nums)

    print("Previous Greater Elements:")
    print(" ".join(map(str, result)))

main()